File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.idx = 0
        self.ans = None
        def dfs(node):
            if not node or self.ans:
                return
            
            dfs(node.left)

            self.idx += 1
            
            if self.idx == k:
                self.ans = node
            
            dfs(node.right)
            return
        

        dfs(root)

        return self.ans.val
        

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        if (!root) return true;
        const dfs = (node) => {
            let left = node.left ? dfs(node.left): 0;
            let right = node.right ? dfs(node.right): 0;
            if (left === -1 || right === -1 || Math.abs(left-right) > 1) return -1;
            return  Math.max(left, right)+1;
        }
        return dfs(root) !== -1
    }
}
